### ISSUE TITLE
Routing tasks to different assistant pools

### ISSUE THREAD

**Issue by davemt** [spotify/luigi#1546]
We are interested in starting to use the assistant functionality, however we need a way to have different pools of assistants that pick up only certain tasks. For example, we want to dedicate a pool of X number of assistant processes to work on a background pipeline, and then have other pool(s) dedicated to other efforts, such as a user running an ad-hoc job.

It is possible the existing `resources` feature could be leveraged and just have a single pool. It is not as flexible though, for example does not allow customization such as different configuration.

Maybe something like an `--assistant-pool` argument which would be specified on both the submitting worker and the assistant side?

One thing that is a little odd is that when tasks overlap from two submissions to different pools, the task actually could be run in either pool. I think this is OK though -- it is no different from the caveats standard worker case.

**Comment by erikbern** [comment]
makes sense. it's kind of like how queue systems work (beanstalk, gearman, rabbitmq) iirc.

**Comment by luckyrandom** [comment]
Does it make sense to let `workers` own `resources`, besides the global `resources` pool? So the assistant can be controlled on task level, such as assistant for hadoop jobs, and assistant for db jobs.

**Comment by davemt** [comment]
That would be good to have, but I believe it is a separate feature. If you were to route just based on resources, you can't also use that to route entire workflows to designated assistant pools.

At first I was thinking you could, just by adding another item to `resources` which is the "pool" (and somehow set this for all tasks in flow). The assistant would then be configured to run task-resources `pool_name,hadoop,db`, where `hadoop` and `db` are only other resource types in the workflow. (That would mean "assistant can run tasks that use _any_ of these resources".)

It breaks down when you have two pools though, because if you had `pool_name2,hadoop`, the `hadoop` resource type overlaps and breaks the routing (can go to either assistant). You could make a `hadoop2`, but that would not be ideal and you have issues then with global resource limits.

**Comment by Tarrasch** [comment]
Would it make sense for the task_namespace to be the pool? Once we've consolidated task_namespace and python_module, it should feel natural. I imagine team A have the python_package `foo.*` and team B has `bar.*`. It would then nice if team A  could conveniently run only their tasks (if they so wish) with an assistant specifying only the name of their package. They could then also only ship their source code from their package. Team B can then do whatever they want interdependently and team A wastes not time with configuration.

**Comment by davemt** [comment]
I did think about this as an option. This is a bit similar to what we do in production where we have multiple applications with very different use cases for luigi. We have an App object which facilitates each named application being configured differently. The App also used to do auto-namespacing so that the app would get a namespace equal to its name, but we removed for a small complication b/c it was not being used for anything (other than convenience in scheduler UI). 

One problem I have with namespace for assistant pool is that it overloads a generic functionality that people may be using for something else already. So if they then wanted the assistant pools but already were using namespace to some other effect, it would not allow both. Or if namespace + module were consolidated and for some reason it was desired to run different pools for the same base module. I guess at some point flexibility must be sacrificed though.

Can you explain more about the task_namespace and python_package consolidation?

**Comment by Tarrasch** [comment]
Although I can imagine organizations have started to use name-spaces for some purpose already. I think many still just use the default empty namespace. Since we've not recommended any convention for the task_namespace I think it's mostly a confusing feature as it is today. Hence why I (and Spotify too, last time I checked) wanted them to be merged. There have been some PR/issue where there's been discussion but I can't find it now. :-/

**Comment by mfcabrera** [comment]
Ping @Tarrasch  not sure what to do with this one. Close it?

**Comment by Tarrasch** [comment]
I'll add this to my own todo. :)
